File: libs/go/src/domain/email.go
package domain

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

// emailRegex is a simplified RFC 5322 compliant email regex.
var emailRegex = regexp.MustCompile(`^[a-zA-Z0-9.!#$%&'*+/=?^_` + "`" + `{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$`)

// Email represents a validated email address.
type Email struct {
	value string
}

// NewEmail creates a new Email from a string, validating RFC 5322 compliance.
func NewEmail(value string) (Email, error) {
	normalized := strings.TrimSpace(strings.ToLower(value))
	if normalized == "" {
		return Email{}, fmt.Errorf("email cannot be empty")
	}
	if len(normalized) > 254 {
		return Email{}, fmt.Errorf("email exceeds maximum length of 254 characters")
	}
	if !emailRegex.MatchString(normalized) {
		return Email{}, fmt.Errorf("invalid email format: %s", value)
	}
	return Email{value: normalized}, nil
}

// MustNewEmail creates a new Email, panicking on invalid input.
func MustNewEmail(value string) Email {
	email, err := NewEmail(value)
	if err != nil {
		panic(err)
	}
	return email
}

// String returns the email as a string.
func (e Email) String() string {
	return e.value
}

// LocalPart returns the local part of the email (before @).
func (e Email) LocalPart() string {
	parts := strings.Split(e.value, "@")
	if len(parts) > 0 {
		return parts[0]
	}
	return ""
}

// Domain returns the domain part of the email (after @).
func (e Email) Domain() string {
	parts := strings.Split(e.value, "@")
	if len(parts) > 1 {
		return parts[1]
	}
	return ""
}

// IsEmpty returns true if the email is empty.
func (e Email) IsEmpty() bool {
	return e.value == ""
}

// Equals checks if two emails are equal.
func (e Email) Equals(other Email) bool {
	return e.value == other.value
}

// MarshalJSON implements json.Marshaler.
func (e Email) MarshalJSON() ([]byte, error) {
	return json.Marshal(e.value)
}

// UnmarshalJSON implements json.Unmarshaler.
func (e *Email) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	email, err := NewEmail(s)
	if err != nil {
		return err
	}
	*e = email
	return nil
}

// MarshalText implements encoding.TextMarshaler.
func (e Email) MarshalText() ([]byte, error) {
	return []byte(e.value), nil
}

// UnmarshalText implements encoding.TextUnmarshaler.
func (e *Email) UnmarshalText(data []byte) error {
	email, err := NewEmail(string(data))
	if err != nil {
		return err
	}
	*e = email
	return nil
}

File: libs/elixir/apps/auth_platform/lib/auth_platform/domain/email.ex
defmodule AuthPlatform.Domain.Email do
  @moduledoc """
  RFC 5322 compliant email address value object.

  Email addresses are normalized to lowercase and validated against RFC 5322 format.

  ## Usage

      alias AuthPlatform.Domain.Email

      # Creating emails
      {:ok, email} = Email.new("User@Example.com")
      email.value  # "user@example.com"

      # Bang version raises on invalid input
      email = Email.new!("user@example.com")

      # String conversion
      to_string(email)  # "user@example.com"

      # JSON encoding
      Jason.encode!(email)  # "\\"user@example.com\\""

  """

  @type t :: %__MODULE__{value: String.t()}

  @enforce_keys [:value]
  defstruct [:value]

  # RFC 5322 compliant email regex (simplified but practical)
  @email_regex ~r/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/

  @doc """
  Creates a new Email from a string.

  The email is validated against RFC 5322 format and normalized to lowercase.

  ## Examples

      iex> AuthPlatform.Domain.Email.new("User@Example.com")
      {:ok, %AuthPlatform.Domain.Email{value: "user@example.com"}}

      iex> AuthPlatform.Domain.Email.new("invalid")
      {:error, "invalid email format"}

      iex> AuthPlatform.Domain.Email.new(123)
      {:error, "email must be a string"}

  """
  @spec new(String.t()) :: {:ok, t()} | {:error, String.t()}
  def new(value) when is_binary(value) do
    trimmed = String.trim(value)

    if Regex.match?(@email_regex, trimmed) do
      {:ok, %__MODULE__{value: String.downcase(trimmed)}}
    else
      {:error, "invalid email format"}
    end
  end

  def new(_), do: {:error, "email must be a string"}

  @doc """
  Creates a new Email from a string, raising on invalid input.

  ## Examples

      iex> AuthPlatform.Domain.Email.new!("user@example.com")
      %AuthPlatform.Domain.Email{value: "user@example.com"}

  ## Raises

      AuthPlatform.Domain.Email.new!("invalid")
      # ** (ArgumentError) invalid email format

  """
  @spec new!(String.t()) :: t()
  def new!(value) do
    case new(value) do
      {:ok, email} -> email
      {:error, reason} -> raise ArgumentError, reason
    end
  end

  @doc """
  Returns the email value as a string.

  ## Examples

      iex> email = AuthPlatform.Domain.Email.new!("user@example.com")
      iex> AuthPlatform.Domain.Email.to_string(email)
      "user@example.com"

  """
  @spec to_string(t()) :: String.t()
  def to_string(%__MODULE__{value: value}), do: value

  @doc """
  Returns the local part of the email (before @).

  ## Examples

      iex> email = AuthPlatform.Domain.Email.new!("user@example.com")
      iex> AuthPlatform.Domain.Email.local_part(email)
      "user"

  """
  @spec local_part(t()) :: String.t()
  def local_part(%__MODULE__{value: value}) do
    [local | _] = String.split(value, "@")
    local
  end

  @doc """
  Returns the domain part of the email (after @).

  ## Examples

      iex> email = AuthPlatform.Domain.Email.new!("user@example.com")
      iex> AuthPlatform.Domain.Email.domain(email)
      "example.com"

  """
  @spec domain(t()) :: String.t()
  def domain(%__MODULE__{value: value}) do
    [_, domain] = String.split(value, "@")
    domain
  end

  defimpl String.Chars do
    def to_string(%{value: value}), do: value
  end

  defimpl Jason.Encoder do
    def encode(%{value: value}, opts), do: Jason.Encode.string(value, opts)
  end
end
